### PR TITLE
Laboratory work #4, Alisa Karpova - 22FPL2 

### PR DESCRIPTION
**Pull request by AlisaKarpova**

### CODE REVIEW

**Review by marina-kaz — COMMENTED** [review]
Strong start! Go on :)

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -98,7 +131,33 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+            if not next_tokens:
+                break
+            probable_words = []
+            probability = 0
+            for word, value in next_tokens.items():
+                probability += value
+                probable_words.append((word, value))
+            if probability >= self._p_value:
+                break
+            sorted_words = sorted(probable_words, key=lambda x: (x[1], x[0]), reverse=True)
```
[the remark this answers is not in the document]
Comment (AlisaKarpova) [reply lab_4_fill_words_by_ngrams/main.py:153]: I was trying to sort the list by the most likely tokens

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -98,7 +131,33 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+            if next_tokens is None:
+                raise ValueError
+            if not next_tokens:
+                break
+            probable_words = []
+            probability = 0
+            for word, value in next_tokens.items():
+                probability += value
+                probable_words.append((word, value))
+            if probability >= self._p_value:
```
[the remark this answers is not in the document]
Comment (AlisaKarpova) [reply lab_4_fill_words_by_ngrams/main.py]: I tried to change it

File: lab_4_fill_words_by_ngrams/main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlisaKarpova) [reply lab_4_fill_words_by_ngrams/main.py]: I tried to change it

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -98,6 +135,32 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
                 or if sequence has inappropriate length,
                 or if methods used return None.
         """
+        if not (isinstance(seq_len, int) and seq_len > 0
+                and isinstance(prompt, str) and prompt):
+            raise ValueError
+        encoded_text = self._word_processor.encode(prompt)
+        if not encoded_text:
+            raise ValueError
+        for i in range(seq_len):
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:144]: if the variable is not used anywhere, name it `_`

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -98,6 +135,32 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+            next_tokens = self._model.generate_next_token(encoded_text)
+            if next_tokens is None:
+                raise ValueError
+            if not next_tokens:
+                break
+            sorted_dict = dict(sorted(list(next_tokens.items()), key=lambda x: (x[1], x[0]), reverse=True))
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:150]: dictionaries do not guarantee the order

File: lab_4_fill_words_by_ngrams/main.py [hunk 5]
```diff
@@ -125,6 +191,12 @@ def get_conversion_generator_type(self, generator_type: int) -> str:  # type: ig
         Returns:
             (str): Name of the generator.
         """
+        if generator_type == 0:
+            return 'Greedy Generator'
+        if generator_type == 1:
+            return 'Top-P Generator'
+        if generator_type == 2:
+            return 'Beam Search Generator'
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:199]: consider implementing it as a dictionary

**Review by marina-kaz — COMMENTED** [review]
Good implementations, a lot of progress!

File: lab_4_fill_words_by_ngrams/main.py (hunk 5 above)
Comment (AlisaKarpova) [reply lab_4_fill_words_by_ngrams/main.py:199]: Done

File: lab_4_fill_words_by_ngrams/main.py (hunk 3 above)
Comment (AlisaKarpova) [reply lab_4_fill_words_by_ngrams/main.py:144]: Fixed

File: lab_4_fill_words_by_ngrams/main.py [hunk 6]
```diff
@@ -28,6 +32,19 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not (isinstance(text, str) and text):
+            raise ValueError
+        punctuation_to_replace = '.!?'
+        tokens = []
+        for word in text.lower().split():
+            if word[-1] in punctuation_to_replace:
+                tokens.extend([word[:-1], self._end_of_word_token])
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:41]: no cleaning for chunks with punctuation ending?

File: lab_4_fill_words_by_ngrams/main.py [hunk 7]
```diff
@@ -39,6 +56,12 @@ def _put(self, element: str) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not (isinstance(element, str) and element):
+            raise ValueError
+        if element in self._storage:
+            return None
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:62]: consider only using just one `return None` statement

File: lab_4_fill_words_by_ngrams/main.py [hunk 8]
```diff
@@ -98,7 +136,35 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+            if next_tokens is None:
+                raise ValueError
+            if not next_tokens:
+                break
+            candidates = []
+            for word, value in next_tokens.items():
+                candidates.append((word, value))
+            sorted_candidates = sorted(candidates, key=lambda x: (x[1], x[0]), reverse=True)
+            probability = 0
+            possible_tokens = tuple()
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:156]: why not making it a list if it needs to be changeable?

File: lab_4_fill_words_by_ngrams/main.py [hunk 9]
```diff
@@ -296,6 +430,16 @@ def assess_exam(self, answers: dict[str, str]) -> float:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not (isinstance(answers, dict) and answers):
+            raise ValueError
+        correct = 0
+        total = len(answers)
+        correct_answers = list(self._questions_and_answers.values())
+        answers_list = answers.values()
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:438]: what if the order is mismatched?

**Review by marina-kaz — COMMENTED** [review]
Good! Keep going

**Comment by marina-kaz** [comment]
I think you should try for mark 10

File: lab_4_fill_words_by_ngrams/main.py (hunk 8 above)
Comment (AlisaKarpova) [reply lab_4_fill_words_by_ngrams/main.py:156]: Fixed

File: lab_4_fill_words_by_ngrams/main.py (hunk 7 above)
Comment (AlisaKarpova) [reply lab_4_fill_words_by_ngrams/main.py:62]: I tried to change it

File: lab_4_fill_words_by_ngrams/main.py (hunk 6 above)
Comment (AlisaKarpova) [reply lab_4_fill_words_by_ngrams/main.py]: I tried to fix it

File: lab_4_fill_words_by_ngrams/main.py (hunk 9 above)
Comment (AlisaKarpova) [reply lab_4_fill_words_by_ngrams/main.py]: I changed it

**Review by marina-kaz — COMMENTED** [review]
Good job, great implementation!
If you want to get 10, don't forget to change target score.

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**